### PR TITLE
14 - Catalog

### PR DESCRIPTION
**Pull request by irenlian**
# Changes proposed

Catalog with renters ads was added.

The catalog uses scrolling to load more items. And adjusts items number to load according to the number of items is in row which depends on screen width.

The current basic step is 20 items.

## Demo
Screenshot Before:

Screenshot 2020-08-10 at 23 29 32

Screenshot After:

Screenshot 2020-08-10 at 23 27 15

## New components

Catalog and ListingItem

## Caveats

No

### CODE REVIEW

File: src/components/Catalog/Catalog.tsx [hunk 1]
```diff
@@ -1,43 +1,94 @@
...
+  // Update alignment of items avoiding stretching the last item
+  useEffect(() => {
+    const catalogWidth: number = catalogRef.current?.clientWidth || 0;
+    const itemWidth = firstItemRef.current?.clientWidth || 1;
+
+    const numberInRow = Math.floor(catalogWidth / itemWidth);
+    const hasItemsInLastRow = numberInRow && Math.floor(adsList.length % numberInRow);
+    const plugs = hasItemsInLastRow === 0 ? 0 : numberInRow - hasItemsInLastRow;
+    setPlugsNumber(plugs);
+  });
```
Comment (dimalian) [inline src/components/Catalog/Catalog.tsx]: You should set dependencies at useEffect

File: src/components/Catalog/Catalog.tsx [hunk 2]
```diff
@@ -1,43 +1,94 @@
...
-        {arrayWithAllLanguages.map(language => (
-          <option value={language} key={language}>{language}</option>
-        ))}
-      </select>
-      <span>Total number of renters ads: {adsList.length}</span>
-      <button type="button" onClick={(): void => { addRenter({ id: '1', uuid: '2' }); }}>Add</button>
-    </div>
+    <Box display="flex" flexDirection="column" alignItems="center" height="100%">
+      {!!adsList.length && (
+        <div id="catalog-page" ref={catalogRef}>
```
Comment (dimalian) [inline src/components/Catalog/Catalog.tsx]: I think, you should use a Box instead a div

File: src/components/Catalog/Catalog.tsx [hunk 3]
```diff
@@ -1,43 +1,94 @@
...
+  // First download
+  useEffect(loadMore, []);
+  // Subsequent downloads
+  useInfiniteScroll({ onLoadMore: loadMore });
+
+  // Elements are used to count the number of items in a row
+  const firstItemRef = useRef<HTMLElement>(null);
+  const catalogRef = useRef<HTMLDivElement>(null);
+
+  const plug = (index: number): JSX.Element => <div className="item-container" key={index} />;
```
Comment (dimalian) [inline src/components/Catalog/Catalog.tsx]: I think, you should use a Box instead a div

File: src/components/Catalog/ListingItem.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      <Box display="flex" alignItems="center">
+        <Typography variant="button">
+          {item.minBudget} {item.currency}
+        </Typography>
+      </Box>
+    </Box>
+  </Card>
+));
+
+export default ListingItem;
```
Comment (dimalian) [inline src/components/Catalog/ListingItem.tsx:43]: Perhaps, you can use standart ListItem from material

File: src/components/Sidebar/Sidebar.tsx [hunk 5]
```diff
@@ -80,19 +96,32 @@ const Sidebar: React.FC = () => {
...
-            <MenuIcon />
-          </IconButton>
-          <Typography variant="h6" noWrap>
-            <FormattedMessage {...title} />
-          </Typography>
+        <Toolbar className="toolbar">
+          <Box display="flex" alignItems="center">
+            <IconButton
+              color="inherit"
+              aria-label="open drawer"
```
Comment (dimalian) [inline src/components/Sidebar/Sidebar.tsx:103]: Maybe, we need move it to intl if it displayable text

File: src/hooks/useInfiniteScroll.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  onLoadMore: Function;
+  disabled?: boolean;
+  selectorId?: string;
+};
+
+export const useInfiniteScroll = ({ onLoadMore, disabled = false, selectorId = 'content' }: Types) => {
+  useEffect(() => {
+    const element = document.getElementById(selectorId) || document.documentElement;
+    const handleScroll = () => {
+      const isPageEnd = element.offsetHeight + element.scrollTop >= element.scrollHeight - 40;
```
Comment (dimalian) [inline src/hooks/useInfiniteScroll.ts]: Maybe, 40 must be constant

File: src/scss/_layout.scss [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 .root {
   display: flex;
+  //min-height: 100vh;
```
Comment (dimalian) [inline src/scss/_layout.scss]: clean

File: src/components/Catalog/Catalog.tsx (hunk 1 above)
Comment (irenlian) [reply src/components/Catalog/Catalog.tsx]: You're right, I' going to move `numberInRow` and use it as a dependency.

File: src/components/Catalog/ListingItem.tsx (hunk 4 above)
Comment (irenlian) [reply src/components/Catalog/ListingItem.tsx:43]: ListItem has other styles, so the custom component is more useful.

File: src/components/Sidebar/Sidebar.tsx (hunk 5 above)
Comment (irenlian) [reply src/components/Sidebar/Sidebar.tsx:103]: It isn't displayable so we can miss its translation

File: src/hooks/useInfiniteScroll.ts (hunk 6 above)
Comment (irenlian) [reply src/hooks/useInfiniteScroll.ts]: I have changed the calculation.

**Review by dimalian — APPROVED** [review]

**Pull request merged**